src/ingest.rs

```diff
@@ -0,0 +1,806 @@
+                        if !self.opts.failed_move_fall_back_to_copy {
+                            return Err(err.into());
+                        }
+                    }
+                }
+            }
+            // copy file
+            match fs::copy(&file.input_path, &out_path) {
+                Ok(_) => {
+                    file.moved_or_copied = true;
+                }
+                Err(err) => {
+                    warn!(
+                        "[ingest task]: failed to copy file {} to db dir. {}",
+                        file.input_path, err
+                    );
+                    return Err(err.into());
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn verify(&mut self) -> Result<()> {
+        use ChecksumVerificationMode::*;
+
+        for file in self.files.iter_mut() {
+            let table = Table::open(
+                &table::new_filename(file.id, &self.core.opts.dir),
+                build_table_options(&self.core.opts),
+            )?;
+            // checksum has been checked when open in [`OnTableRead`] or [`OnTableAndBlockRead`] mode
+            // avoid double check
+            if self.opts.verify_checksum
+                && !matches!(
+                    self.core.opts.checksum_mode,
+                    OnTableRead | OnTableAndBlockRead
+                )
+            {
+                table.verify_checksum()?;
+            }
+            file.table = Some(table);
+        }
+        Ok(())
+    }
+
+    fn assign_version(&mut self) {
+        let user_key_cmp = FixedLengthSuffixComparator::new(0);
+        // all tables opened, sort by range
+        self.files.sort_unstable_by(|x, y| {
+            user_key_cmp.compare_key(
+                user_key(x.table.as_ref().unwrap().smallest()),
+                user_key(y.table.as_ref().unwrap().smallest()),
+            )
+        });
+        // check overlap
+        if self.files.len() > 1 {
+            for i in 0..(self.files.len() - 1) {
+                if matches!(
+                    user_key_cmp.compare_key(
+                        user_key(self.files[i].table.as_ref().unwrap().biggest()),
+                        user_key(self.files[i + 1].table.as_ref().unwrap().smallest()),
+                    ),
+                    Ordering::Equal | Ordering::Greater
+                ) {
+                    self.overlap = true;
+                    break;
+                }
+            }
+        }
+        if self.overlap {
+            // files overlap with each other, assign commit_ts for each file
+            for file in self.files.iter_mut() {
+                let range = RangeInclusive::new(
+                    Bytes::copy_from_slice(user_key(file.table.as_ref().unwrap().smallest())),
+                    Bytes::copy_from_slice(user_key(file.table.as_ref().unwrap().biggest())),
+                );
+                let version = self.core.orc.new_ingest_commit_ts(vec![range], &self.opts);
```

Comment:
Can we only get commit timestamp once for all external files?

Reply:
I'd like to. But for files that overlap with each other, there might be same key. And if we get one commit_ts for all these files, there will be same key with same version in different SST which breaks the protection.